action/protocol/staking/bucket_validation.go

```diff
@@ -77,24 +77,27 @@ func validateBucketSelfStake(featureCtx protocol.FeatureCtx, csm CandidateStateM
 }
 
 func validateBucketEndorsement(esm *EndorsementStateManager, bucket *VoteBucket, isEndorsed bool, height uint64) ReceiptError {
+	var status EndorsementStatus
+	if isEndorsed {
+		status = Endorsed | UnEndorsing
+	} else {
+		status = WithoutEndorsement | EndorseExpired
+	}
+	return validateBucketEndorsementByStatus(esm, bucket, status, height)
+}
+
+func validateBucketEndorsementByStatus(esm *EndorsementStateManager, bucket *VoteBucket, status EndorsementStatus, height uint64) ReceiptError {
```

Comment:
Efficiency is not the first thing to consider in this case, thus this refactor is not suggested.

Reply:
reverted